File: packages/exchange-adapter/src/exchanges/okx/PublicAdapter.ts
import { RestClient } from 'okx-api'
import type { AxiosRequestConfig } from 'axios'
import type {
  Exchange,
  TradeType,
  Result,
  SymbolInfo,
  Ticker,
  OrderBook,
  AdapterOptions
} from '../../core/types'
import { SymbolStatus } from '../../core/types'
import { Ok, Err, getDecimalPlaces, wrapAsync, createProxyAgent } from '../../core/utils'
import { BasePublicAdapter } from '../../core/BasePublicAdapter'
import { ErrorCodes } from '../../core/errorCodes'
import {
  unifiedToOkx,
  okxToUnified,
  getOkxInstType,
} from './utils'

// OKX API response types
interface OkxInstrumentResponse {
  instId: string
  baseCcy: string
  quoteCcy: string
  tickSz: string
  lotSz: string
  minSz: string
  maxLmtSz: string
  state: string
  ctVal?: string
  lever?: string
}

interface OkxTickerResponse {
  instId: string
  last: string
  high24h: string
  low24h: string
  vol24h: string
  volCcy24h: string
  ts: string
}

interface OkxOrderBookResponse {
  asks: [string, string, string, string][]
  bids: [string, string, string, string][]
  ts: string
}

interface OkxMarkPriceResponse {
  markPx: string
}

/**
 * OKX 公共 API 适配器
 */
export class OkxPublicAdapter extends BasePublicAdapter {
  readonly exchange: Exchange = 'okx'

  protected client: RestClient

  constructor(options?: AdapterOptions) {
    super()
    const requestOptions: AxiosRequestConfig = {}

    if (options?.httpsProxy || options?.socksProxy) {
      const agent = createProxyAgent(options)
      requestOptions.httpAgent = agent
      requestOptions.httpsAgent = agent
    }

    this.client = new RestClient({}, requestOptions)
  }

  /**
   * 获取交易对信息
   */
  async getSymbolInfo(symbol: string, tradeType: TradeType): Promise<Result<SymbolInfo>> {
    // 先检查缓存
    const cached = this.getCachedSymbol(symbol, tradeType)
    if (cached) {
      return Ok(cached)
    }

    const instId = unifiedToOkx(symbol, tradeType)
    const instType = getOkxInstType(tradeType)
    const result = await wrapAsync<OkxInstrumentResponse[]>(
      () => this.client.getInstruments({ instType, instId }),
      'GET_SYMBOL_INFO_ERROR'
    )

    if (!result.ok) {
      return Err(result.error)
    }

    const data = result.data

    if (!data || data.length === 0) {
      return Err({
        code: ErrorCodes.SYMBOL_NOT_FOUND,
        message: `Symbol ${symbol} not found`
      })
    }

    const inst = data[0]
    const symbolInfo = this.transformSymbolInfo(inst, tradeType)

    // 设置缓存
    this.setCachedSymbol(symbol, tradeType, symbolInfo)

    return Ok(symbolInfo)
  }

  /**
   * 获取所有交易对信息
   */
  async getAllSymbols(tradeType: TradeType): Promise<Result<SymbolInfo[]>> {
    const instType = getOkxInstType(tradeType)

    const result = await wrapAsync<OkxInstrumentResponse[]>(
      () => this.client.getInstruments({ instType }),
      'GET_ALL_SYMBOLS_ERROR'
    )

    if (!result.ok) {
      return Err(result.error)
    }

    const symbols = result.data.map(inst => this.transformSymbolInfo(inst, tradeType))

    // 批量设置缓存
    this.setCachedSymbols(tradeType, symbols)

    return Ok(symbols)
  }

  /**
   * 获取当前价格
   */
  async getPrice(symbol: string, tradeType: TradeType): Promise<Result<string>> {
    const instId = unifiedToOkx(symbol, tradeType)

    const result = await wrapAsync<OkxTickerResponse[]>(
      () => this.client.getTicker({ instId }),
      'GET_PRICE_ERROR'
    )

    if (!result.ok) {
      return Err(result.error)
    }

    const data = result.data

    if (!data || data.length === 0) {
      return Err({
        code: ErrorCodes.PRICE_NOT_FOUND,
        message: `Price for ${symbol} not found`
      })
    }

    return Ok(data[0].last)
  }

  /**
   * 获取 Ticker
   */
  async getTicker(symbol: string, tradeType: TradeType): Promise<Result<Ticker>> {
    const instId = unifiedToOkx(symbol, tradeType)

    const result = await wrapAsync<OkxTickerResponse[]>(
      () => this.client.getTicker({ instId }),
      'GET_TICKER_ERROR'
    )

    if (!result.ok) {
      return Err(result.error)
    }

    const data = result.data

    if (!data || data.length === 0) {
      return Err({
        code: ErrorCodes.TICKER_NOT_FOUND,
        message: `Ticker for ${symbol} not found`
      })
    }

    const ticker = data[0]
    // OKX 不直接返回涨跌幅，需要从其他接口获取或计算
    // 这里暂时返回 0
    return Ok({
      symbol,
      last: ticker.last,
      high: ticker.high24h,
      low: ticker.low24h,
      volume: ticker.vol24h,
      quoteVolume: ticker.volCcy24h,
      timestamp: parseInt(ticker.ts)
    })
  }

  /**
   * 获取深度数据
   */
  async getOrderBook(
    symbol: string,
    tradeType: TradeType,
    limit: number = 20
  ): Promise<Result<OrderBook>> {
    const instId = unifiedToOkx(symbol, tradeType)

    const result = await wrapAsync<OkxOrderBookResponse[]>(
      () => this.client.getOrderBook({ instId, sz: String(limit) }),
      'GET_ORDERBOOK_ERROR'
    )

    if (!result.ok) {
      return Err(result.error)
    }

    const data = result.data

    if (!data || data.length === 0) {
      return Err({
        code: ErrorCodes.ORDERBOOK_NOT_FOUND,
        message: `OrderBook for ${symbol} not found`
      })
    }

    const orderBook = data[0]
    return Ok({
      symbol,
      asks: orderBook.asks.map(([price, qty]) => [price, qty] as [string, string]),
      bids: orderBook.bids.map(([price, qty]) => [price, qty] as [string, string]),
      timestamp: parseInt(orderBook.ts)
    })
  }

  /**
   * 获取标记价格 (合约)
   */
  async getMarkPrice(symbol: string, tradeType: TradeType): Promise<Result<string>> {
    if (tradeType === 'spot') {
      // 现货没有标记价格，返回最新价
      return this.getPrice(symbol, tradeType)
    }

    const instId = unifiedToOkx(symbol, tradeType)

    const result = await wrapAsync<OkxMarkPriceResponse[]>(
      () => this.client.getMarkPrice({ instId }),
      'GET_MARK_PRICE_ERROR'
    )

    if (!result.ok) {
      return Err(result.error)
    }

    const data = result.data

    if (!data || data.length === 0) {
      return Err({
        code: ErrorCodes.MARK_PRICE_NOT_FOUND,
        message: `Mark price for ${symbol} not found`
      })
    }

    return Ok(data[0].markPx)
  }

  /**
   * 统一格式 -> OKX 原始格式
   */
  toRawSymbol(symbol: string, tradeType: TradeType): string {
    return unifiedToOkx(symbol, tradeType)
  }

  /**
   * OKX 原始格式 -> 统一格式
   */
  fromRawSymbol(rawSymbol: string, _tradeType: TradeType): string {
    const { symbol } = okxToUnified(rawSymbol)
    return symbol
  }

  // ============================================================================
  // 私有方法
  // ============================================================================

  /**
   * 转换 OKX 的 instrument 数据为统一格式
   */
  private transformSymbolInfo(
    inst: OkxInstrumentResponse,
    tradeType: TradeType
  ): SymbolInfo {
    // 从 instId 解析出统一 symbol
    // BTC-USDT -> BTC-USDT
    // BTC-USDT-SWAP -> BTC-USDT
    const parts = inst.instId.split('-')
    const unifiedSymbol = `${parts[0]}-${parts[1]}`

    return {
      symbol: unifiedSymbol,
      rawSymbol: inst.instId,
      baseCurrency: inst.baseCcy || parts[0],
      quoteCurrency: inst.quoteCcy || parts[1],
      tradeType,
      tickSize: inst.tickSz,
      stepSize: inst.lotSz,
      minQty: inst.minSz,
      maxQty: inst.maxLmtSz,
      quantityPrecision: getDecimalPlaces(inst.lotSz),
      pricePrecision: getDecimalPlaces(inst.tickSz),
      status: inst.state === 'live' ? SymbolStatus.Enabled : SymbolStatus.Disabled,
      contractValue: Number(inst.ctVal) || 0,
      maxLeverage: inst.lever ? parseInt(inst.lever) : 0
    }
  }
}

File: pkg/exchange-adapter/exchanges/okx/public_adapter.go
package okx

import (
	"context"
	"encoding/json"
	"fmt"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/pkg/exchange-adapter/core"
	"github.com/pkg/exchange-adapter/marketdata"
	okxapi "github.com/pkg/okx-api"
	okxtypes "github.com/pkg/okx-api/types"
	okxrest "github.com/pkg/okx-api/types/rest"
	okxutil "github.com/pkg/okx-api/util"
)

type PublicAdapter struct {
	core.BasePublicAdapter

	client  *okxapi.RestClient
	initErr error
}

func NewPublicAdapter(options *core.AdapterOptions) *PublicAdapter {
	demonet := true
	if options != nil && options.Demonet != nil {
		demonet = *options.Demonet
	}

	proxyURL := ""
	socksProxyURL := ""
	if options != nil {
		proxyURL = options.HTTPSProxy
		socksProxyURL = options.SOCKSProxy
	}

	client, err := okxapi.NewRestClient(okxrest.RestClientOptions{
		Market:          okxtypes.APIMarketGLOBAL,
		DemoTrading:     demonet,
		ParseExceptions: true,
		Proxy:           proxyURL,
		SocksProxy:      socksProxyURL,
	})

	return &PublicAdapter{
		BasePublicAdapter: core.NewBasePublicAdapter(core.ExchangeOKX, time.Hour),
		client:            client,
		initErr:           err,
	}
}

func (a *PublicAdapter) ToRawSymbol(symbol string, tradeType core.TradeType) string {
	return unifiedToRawSymbol(symbol, tradeType)
}

func (a *PublicAdapter) FromRawSymbol(rawSymbol string, _ core.TradeType) string {
	return rawToUnifiedSymbol(rawSymbol)
}

func (a *PublicAdapter) GetSymbolInfo(ctx context.Context, symbol string, tradeType core.TradeType) core.Result[core.SymbolInfo] {
	if cached, ok := a.GetCachedSymbol(symbol, tradeType); ok {
		return core.Ok(cached)
	}

	all := a.GetAllSymbols(ctx, tradeType)
	if !all.Ok {
		return core.Err[core.SymbolInfo](*all.Error)
	}

	for _, info := range all.Data {
		if info.Symbol == symbol {
			return core.Ok(info)
		}
	}

	return core.Err[core.SymbolInfo](core.ErrorInfo{
		Code:    core.ErrorSymbolNotFound,
		Message: fmt.Sprintf("symbol not found: %s (%s)", symbol, tradeType),
	})
}

type okxInstrument struct {
	InstID   string `json:"instId"`
	BaseCcy  string `json:"baseCcy"`
	QuoteCcy string `json:"quoteCcy"`
	TickSz   string `json:"tickSz"`
	LotSz    string `json:"lotSz"`
	MinSz    string `json:"minSz"`
	MaxLmtSz string `json:"maxLmtSz"`
	State    string `json:"state"`
	CtVal    string `json:"ctVal,omitempty"`
	Lever    string `json:"lever,omitempty"`
}

type getInstrumentsParams struct {
	InstType string `json:"instType"`
	InstID   string `json:"instId,omitempty"`
}

func (a *PublicAdapter) GetAllSymbols(ctx context.Context, tradeType core.TradeType) core.Result[[]core.SymbolInfo] {
	if a.initErr != nil {
		return core.Err[[]core.SymbolInfo](core.ErrorInfo{Code: "INIT_ERROR", Message: a.initErr.Error(), Raw: a.initErr})
	}

	raw, err := a.client.GetInstruments(ctx, getInstrumentsParams{InstType: tradeTypeToInstType(tradeType)})
	if err != nil {
		return core.Err[[]core.SymbolInfo](wrapAPIError("GET_ALL_SYMBOLS_ERROR", err))
	}

	var insts []okxInstrument
	if uerr := json.Unmarshal(raw, &insts); uerr != nil {
		return core.Err[[]core.SymbolInfo](core.ErrorInfo{Code: "PARSE_ERROR", Message: uerr.Error(), Raw: string(raw)})
	}

	// OKX delivery futures have multiple expiries per base/quote, which would collide in our symbol cache.
	// Deduplicate by selecting the earliest expiry (lexicographically) per unified symbol.
	if tradeType == core.TradeTypeDelivery {
		insts = dedupDeliveryInstruments(insts)
	}

	out := make([]core.SymbolInfo, 0, len(insts))
	for _, inst := range insts {
		info := transformSymbolInfo(inst, tradeType)
		out = append(out, info)
	}

	a.SetCachedSymbols(tradeType, out)
	return core.Ok(out)
}

func dedupDeliveryInstruments(insts []okxInstrument) []okxInstrument {
	best := make(map[string]okxInstrument)
	for _, inst := range insts {
		unified := rawToUnifiedSymbol(inst.InstID)
		prev, ok := best[unified]
		if !ok {
			best[unified] = inst
			continue
		}

		prevDate := deliveryExpiry(prev.InstID)
		thisDate := deliveryExpiry(inst.InstID)

		// If we can compare expiry, pick the earliest; otherwise keep the existing one.
		if thisDate != "" && (prevDate == "" || thisDate < prevDate) {
			best[unified] = inst
		}
	}

	keys := make([]string, 0, len(best))
	for k := range best {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	out := make([]okxInstrument, 0, len(best))
	for _, k := range keys {
		out = append(out, best[k])
	}
	return out
}

func deliveryExpiry(instID string) string {
	parts := strings.Split(instID, "-")
	if len(parts) == 3 && isDigits(parts[2]) {
		return parts[2]
	}
	return ""
}

func transformSymbolInfo(inst okxInstrument, tradeType core.TradeType) core.SymbolInfo {
	parts := strings.Split(inst.InstID, "-")
	unifiedSymbol := rawToUnifiedSymbol(inst.InstID)

	base := inst.BaseCcy
	quote := inst.QuoteCcy
	if base == "" && len(parts) > 0 {
		base = parts[0]
	}
	if quote == "" && len(parts) > 1 {
		quote = parts[1]
	}

	status := core.SymbolStatusDisabled
	if strings.EqualFold(inst.State, "live") {
		status = core.SymbolStatusEnabled
	}

	var contractValue *float64
	if inst.CtVal != "" {
		if f, err := strconv.ParseFloat(inst.CtVal, 64); err == nil && f > 0 {
			contractValue = &f
		}
	}

	var maxLeverage *float64
	if inst.Lever != "" {
		if f, err := strconv.ParseFloat(inst.Lever, 64); err == nil && f > 0 {
			maxLeverage = &f
		}
	}

	return core.SymbolInfo{
		Symbol:            unifiedSymbol,
		RawSymbol:         inst.InstID,
		BaseCurrency:      base,
		QuoteCurrency:     quote,
		TradeType:         tradeType,
		TickSize:          inst.TickSz,
		StepSize:          inst.LotSz,
		MinQty:            inst.MinSz,
		MaxQty:            inst.MaxLmtSz,
		QuantityPrecision: core.GetDecimalPlaces(inst.LotSz),
		PricePrecision:    core.GetDecimalPlaces(inst.TickSz),
		Status:            status,
		ContractValue:     contractValue,
		MaxLeverage:       maxLeverage,
	}
}

type okxTicker struct {
	Last      string `json:"last"`
	High24h   string `json:"high24h"`
	Low24h    string `json:"low24h"`
	Vol24h    string `json:"vol24h"`
	VolCcy24h string `json:"volCcy24h"`
	Ts        string `json:"ts"`
}

type getTickerParams struct {
	InstID string `json:"instId"`
}

func (a *PublicAdapter) GetPrice(ctx context.Context, symbol string, tradeType core.TradeType) core.Result[string] {
	if a.initErr != nil {
		return core.Err[string](core.ErrorInfo{Code: "INIT_ERROR", Message: a.initErr.Error(), Raw: a.initErr})
	}

	instID, err := a.resolveInstID(ctx, symbol, tradeType)
	if err != nil {
		return core.Err[string](core.ErrorInfo{Code: core.ErrorInvalidParams, Message: err.Error(), Raw: err})
	}

	raw, err := a.client.GetTicker(ctx, getTickerParams{InstID: instID})
	if err != nil {
		return core.Err[string](wrapAPIError(core.ErrorPriceNotFound, err))
	}

	var ticks []okxTicker
	if uerr := json.Unmarshal(raw, &ticks); uerr != nil {
		return core.Err[string](core.ErrorInfo{Code: "PARSE_ERROR", Message: uerr.Error(), Raw: string(raw)})
	}
	if len(ticks) == 0 {
		return core.Err[string](core.ErrorInfo{Code: core.ErrorPriceNotFound, Message: "empty ticker response", Raw: string(raw)})
	}
	return core.Ok(ticks[0].Last)
}

func (a *PublicAdapter) GetTicker(ctx context.Context, symbol string, tradeType core.TradeType) core.Result[core.Ticker] {
	if a.initErr != nil {
		return core.Err[core.Ticker](core.ErrorInfo{Code: "INIT_ERROR", Message: a.initErr.Error(), Raw: a.initErr})
	}

	instID, err := a.resolveInstID(ctx, symbol, tradeType)
	if err != nil {
		return core.Err[core.Ticker](core.ErrorInfo{Code: core.ErrorInvalidParams, Message: err.Error(), Raw: err})
	}

	raw, err := a.client.GetTicker(ctx, getTickerParams{InstID: instID})
	if err != nil {
		return core.Err[core.Ticker](wrapAPIError(core.ErrorTickerNotFound, err))
	}

	var ticks []okxTicker
	if uerr := json.Unmarshal(raw, &ticks); uerr != nil {
		return core.Err[core.Ticker](core.ErrorInfo{Code: "PARSE_ERROR", Message: uerr.Error(), Raw: string(raw)})
	}
	if len(ticks) == 0 {
		return core.Err[core.Ticker](core.ErrorInfo{Code: core.ErrorTickerNotFound, Message: "empty ticker response", Raw: string(raw)})
	}

	ts, _ := strconv.ParseInt(ticks[0].Ts, 10, 64)
	return core.Ok(core.Ticker{
		Symbol:      symbol,
		Last:        ticks[0].Last,
		High:        ticks[0].High24h,
		Low:         ticks[0].Low24h,
		Volume:      ticks[0].Vol24h,
		QuoteVolume: ticks[0].VolCcy24h,
		Timestamp:   ts,
		Raw:         json.RawMessage(raw),
	})
}

type okxOrderBook struct {
	Asks [][]string `json:"asks"`
	Bids [][]string `json:"bids"`
	Ts   string     `json:"ts"`
}

type getOrderBookParams struct {
	InstID string `json:"instId"`
	Sz     string `json:"sz,omitempty"`
}

func (a *PublicAdapter) GetOrderBook(ctx context.Context, symbol string, tradeType core.TradeType, limit int) core.Result[core.OrderBook] {
	if a.initErr != nil {
		return core.Err[core.OrderBook](core.ErrorInfo{Code: "INIT_ERROR", Message: a.initErr.Error(), Raw: a.initErr})
	}
	if limit <= 0 {
		limit = 20
	}

	instID, err := a.resolveInstID(ctx, symbol, tradeType)
	if err != nil {
		return core.Err[core.OrderBook](core.ErrorInfo{Code: core.ErrorInvalidParams, Message: err.Error(), Raw: err})
	}

	raw, err := a.client.GetOrderBook(ctx, getOrderBookParams{InstID: instID, Sz: strconv.Itoa(limit)})
	if err != nil {
		return core.Err[core.OrderBook](wrapAPIError(core.ErrorOrderBookNotFound, err))
	}

	var books []okxOrderBook
	if uerr := json.Unmarshal(raw, &books); uerr != nil {
		return core.Err[core.OrderBook](core.ErrorInfo{Code: "PARSE_ERROR", Message: uerr.Error(), Raw: string(raw)})
	}
	if len(books) == 0 {
		return core.Err[core.OrderBook](core.ErrorInfo{Code: core.ErrorOrderBookNotFound, Message: "empty order book response", Raw: string(raw)})
	}

	book := books[0]
	asks := make([][2]string, 0, len(book.Asks))
	for _, a := range book.Asks {
		if len(a) < 2 {
			continue
		}
		asks = append(asks, [2]string{a[0], a[1]})
	}
	bids := make([][2]string, 0, len(book.Bids))
	for _, b := range book.Bids {
		if len(b) < 2 {
			continue
		}
		bids = append(bids, [2]string{b[0], b[1]})
	}

	ts, _ := strconv.ParseInt(book.Ts, 10, 64)
	return core.Ok(core.OrderBook{
		Symbol:    symbol,
		Asks:      asks,
		Bids:      bids,
		Timestamp: ts,
		Raw:       json.RawMessage(raw),
	})
}

type okxMarkPrice struct {
	MarkPx string `json:"markPx"`
}

type getMarkPriceParams struct {
	InstID string `json:"instId"`
}

func (a *PublicAdapter) GetMarkPrice(ctx context.Context, symbol string, tradeType core.TradeType) core.Result[string] {
	if tradeType == core.TradeTypeSpot {
		return a.GetPrice(ctx, symbol, tradeType)
	}
	if a.initErr != nil {
		return core.Err[string](core.ErrorInfo{Code: "INIT_ERROR", Message: a.initErr.Error(), Raw: a.initErr})
	}

	instID, err := a.resolveInstID(ctx, symbol, tradeType)
	if err != nil {
		return core.Err[string](core.ErrorInfo{Code: core.ErrorInvalidParams, Message: err.Error(), Raw: err})
	}

	raw, err := a.client.GetMarkPrice(ctx, getMarkPriceParams{InstID: instID})
	if err != nil {
		return core.Err[string](wrapAPIError(core.ErrorMarkPriceNotFound, err))
	}

	var mp []okxMarkPrice
	if uerr := json.Unmarshal(raw, &mp); uerr != nil {
		return core.Err[string](core.ErrorInfo{Code: "PARSE_ERROR", Message: uerr.Error(), Raw: string(raw)})
	}
	if len(mp) == 0 {
		return core.Err[string](core.ErrorInfo{Code: core.ErrorMarkPriceNotFound, Message: "empty mark price response", Raw: string(raw)})
	}
	return core.Ok(mp[0].MarkPx)
}

func (a *PublicAdapter) resolveInstID(ctx context.Context, symbol string, tradeType core.TradeType) (string, error) {
	// For SPOT/FUTURES, we can derive deterministically.
	if tradeType == core.TradeTypeSpot || tradeType == core.TradeTypeFutures {
		return unifiedToRawSymbol(symbol, tradeType), nil
	}

	// Delivery needs the expiry suffix. If the caller passes an instId-like value, accept it.
	if detectTradeTypeFromInstID(symbol) == core.TradeTypeDelivery {
		return symbol, nil
	}

	info := a.GetSymbolInfo(ctx, symbol, tradeType)
	if info.Ok {
		return info.Data.RawSymbol, nil
	}
	return "", fmt.Errorf("failed to resolve instId for %s (%s)", symbol, tradeType)
}

func wrapAPIError(fallbackCode string, err error) core.ErrorInfo {
	if err == nil {
		return core.ErrorInfo{Code: fallbackCode, Message: "unknown error"}
	}
	if apiErr, ok := err.(*okxutil.APIError); ok {
		code := apiErr.Code
		if code == "" {
			code = fallbackCode
		}
		msg := apiErr.Message
		if msg == "" {
			msg = apiErr.Error()
		}
		return core.ErrorInfo{Code: code, Message: msg, Raw: apiErr}
	}
	return core.ErrorInfo{Code: fallbackCode, Message: err.Error(), Raw: err}
}

type MarketDataRESTClientOptions struct {
	HTTPSProxy  string
	SOCKSProxy  string
	DemoTrading bool
}

// MarketDataRESTClient implements `marketdata.RESTClient` for exchange-sync.
type MarketDataRESTClient struct {
	client *okxapi.RestClient
}

func NewMarketDataRESTClient(opts MarketDataRESTClientOptions) (*MarketDataRESTClient, error) {
	client, err := okxapi.NewRestClient(okxrest.RestClientOptions{
		Market:          okxtypes.APIMarketGLOBAL,
		DemoTrading:     opts.DemoTrading,
		ParseExceptions: true,
		Proxy:           opts.HTTPSProxy,
		SocksProxy:      opts.SOCKSProxy,
	})
	if err != nil {
		return nil, err
	}
	return &MarketDataRESTClient{client: client}, nil
}

func (c *MarketDataRESTClient) Name() marketdata.ExchangeName { return marketdata.OKX }

func (c *MarketDataRESTClient) GetSymbols(ctx context.Context, tradeType marketdata.TradeType) ([]marketdata.SymbolInfo, error) {
	instType := "SPOT"
	if tradeType == marketdata.Futures {
		instType = "SWAP"
	}

	raw, err := c.client.GetInstruments(ctx, getInstrumentsParams{InstType: instType})
	if err != nil {
		return nil, err
	}

	var insts []okxInstrument
	if err := json.Unmarshal(raw, &insts); err != nil {
		return nil, err
	}

	out := make([]marketdata.SymbolInfo, 0, len(insts))
	for _, inst := range insts {
		if !strings.Contains(inst.InstID, "-USDT") {
			continue
		}
		if inst.State != "live" {
			continue
		}

		base := inst.BaseCcy
		quote := inst.QuoteCcy
		if base == "" || quote == "" {
			parts := strings.Split(inst.InstID, "-")
			if base == "" && len(parts) > 0 {
				base = parts[0]
			}
			if quote == "" && len(parts) > 1 {
				quote = parts[1]
			}
		}

		info := marketdata.SymbolInfo{
			Symbol:            marketdata.NormalizeSymbol(marketdata.OKX, inst.InstID, tradeType),
			RawSymbol:         inst.InstID,
			BaseCurrency:      base,
			QuoteCurrency:     quote,
			TradeType:         string(tradeType),
			TickSize:          inst.TickSz,
			StepSize:          inst.LotSz,
			MinQty:            inst.MinSz,
			MaxQty:            inst.MaxLmtSz,
			PricePrecision:    marketdata.CalculatePrecision(inst.TickSz),
			QuantityPrecision: marketdata.CalculatePrecision(inst.LotSz),
			Status:            inst.State,
		}

		if inst.CtVal != "" {
			if v, err := strconv.ParseFloat(inst.CtVal, 64); err == nil {
				info.ContractValue = &v
			}
		}
		if inst.Lever != "" {
			if v, err := strconv.ParseFloat(inst.Lever, 64); err == nil {
				info.MaxLeverage = &v
			}
		}

		out = append(out, info)
	}

	return out, nil
}

type getHistoryCandlesParams struct {
	InstID string `json:"instId"`
	Bar    string `json:"bar,omitempty"`
	After  string `json:"after,omitempty"`
	Before string `json:"before,omitempty"`
	Limit  string `json:"limit,omitempty"`
}

func (c *MarketDataRESTClient) GetCandles(ctx context.Context, symbol string, tradeType marketdata.TradeType, period marketdata.Period, startTime, endTime int64, limit int) ([]marketdata.NormalizedCandle, error) {
	instID := marketdata.ToExchangeSymbol(marketdata.OKX, symbol, tradeType)

	bar := map[marketdata.Period]string{
		marketdata.Period5m:  "5m",
		marketdata.Period15m: "15m",
		marketdata.Period4h:  "4H",
		marketdata.Period1d:  "1D",
	}[period]
	if bar == "" {
		bar = "15m"
	}

	params := getHistoryCandlesParams{
		InstID: instID,
		Bar:    bar,
		Limit:  fmt.Sprintf("%d", limit),
	}
	if endTime > 0 {
		params.After = fmt.Sprintf("%d", endTime)
	}
	if startTime > 0 {
		params.Before = fmt.Sprintf("%d", startTime)
	}

	raw, err := c.client.GetHistoryCandles(ctx, params)
	if err != nil {
		return nil, err
	}

	var rows [][]string
	if err := json.Unmarshal(raw, &rows); err != nil {
		var anyRows [][]any
		if err2 := json.Unmarshal(raw, &anyRows); err2 != nil {
			return nil, err
		}
		rows = make([][]string, 0, len(anyRows))
		for _, r := range anyRows {
			ss := make([]string, 0, len(r))
			for _, v := range r {
				ss = append(ss, fmt.Sprintf("%v", v))
			}
			rows = append(rows, ss)
		}
	}

	normalizedSymbol := marketdata.NormalizeSymbol(marketdata.OKX, instID, tradeType)
	out := make([]marketdata.NormalizedCandle, 0, len(rows))

	for _, k := range rows {
		if len(k) < 6 {
			continue
		}
		ts := core.ParseInt64(k[0])

		out = append(out, marketdata.NormalizedCandle{
			Symbol:       normalizedSymbol,
			Exchange:     string(marketdata.OKX),
			TradeType:    string(tradeType),
			Period:       string(period),
			Timestamp:    ts,
			Open:         core.ParseFloat(k[1]),
			High:         core.ParseFloat(k[2]),
			Low:          core.ParseFloat(k[3]),
			Close:        core.ParseFloat(k[4]),
			Volume:       core.ParseFloat(k[5]),
			BuyVolume:    0,
			SymbolFamily: marketdata.ExtractSymbolFamily(normalizedSymbol),
		})
	}

	// OKX returns newest-first; reverse to ascending time.
	for i, j := 0, len(out)-1; i < j; i, j = i+1, j-1 {
		out[i], out[j] = out[j], out[i]
	}

	return out, nil
}
